src/cluster_legacy.c

```diff
@@ -2596,11 +2596,21 @@ void clusterProcessPingExtensions(clusterMsg *hdr, clusterLink *link) {
         /* We know this will be valid since we validated it ahead of time */
         ext = getNextPingExt(ext);
     }
+
     /* If the node did not send us a hostname extension, assume
      * they don't have an announced hostname. Otherwise, we'll
      * set it now. */
     updateAnnouncedHostname(sender, ext_hostname);
     updateAnnouncedHumanNodename(sender, ext_humannodename);
+    /* If the node did not send us a shard-id extension, it means the sender
+     * does not support it (old version), node->shard_id is randomly generated.
+     * If sender is a replica, set the shard_id to the shard_id of its master.
+     * Otherwise, we'll set it now. */
+    if (ext_shardid == NULL && clusterNodeIsSlave(sender)) {
```

Comment:
I think the logic would be clearer if the `clusterNodeIsSlave` check is removed. 

```suggestion
    if (ext_shardid == NULL) {
```

Furthermore, another nitpick could be creating a helper function to return the primary of any given node, something like the below

```
inline clusterNode* getClusterMasterNode(clusterNode *n) {
    while (n->slaveof != NULL) n=n->slaveof;
    return n;
}
```

then the fix could be written as follows, which conveys the intent better

```
if (ext_shardid == NULL) ext_shardid = getClusterMasterNode(sender)->shard_id;
```

Reply:
thanks for the review!